[lldb-dap] Add feature to remember last non-empty expression.

cmtice opened the pull request:
Update lldb-dap so if the user just presses return, which sends an empty expression, it re-evaluates the most recent non-empty expression/command. Also udpated test to test this case.

cmtice committed:
[lldb-dap] Add feature to remember last non-empty expression.

Update lldb-dap so if the user just presses return, which sends an
empty expression, it re-evaluates the most recent non-empty
expression/command. Also udpated test to test this case.

jeffreytan81 commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -1363,6 +1363,14 @@ void request_evaluate(const llvm::json::Object &request) {
   lldb::SBFrame frame = g_dap.GetLLDBFrame(*arguments);
   std::string expression = GetString(arguments, "expression").str();
   llvm::StringRef context = GetString(arguments, "context");
+  static std::string last_nonempty_expression;
```
Do not use static object. Move to DAP object instead.

cmtice replied:
Done.

cmtice committed:
[lldb] Add feature to remember last non-empty expression

Make last_nonempty_spression part of DAP struct rather than a
static variable.

labath reviewed (commented):
I wrote a fairly long comment on Friday, but I don't see it anymore so, it looks like github has swallowed it. Here's my reconstruction of it:

LLDB commands have the notion of a "repeat command", which can sometimes be more complicated than just running the same string over and over again. This can be e.g. seen with the `memory read` command, which doesn't just print the same memory again -- it actually prints the memory that comes *after* it (a pretty nifty feature actually):
```
(lldb) memory read argv
0x7fffffffd8a8: 49 dc ff ff ff 7f 00 00 00 00 00 00 00 00 00 00  I...............
0x7fffffffd8b8: 6c dc ff ff ff 7f 00 00 ae dc ff ff ff 7f 00 00  l...............
(lldb) 
0x7fffffffd8c8: c4 dc ff ff ff 7f 00 00 cf dc ff ff ff 7f 00 00  ................
0x7fffffffd8d8: 09 dd ff ff ff 7f 00 00 18 dd ff ff ff 7f 00 00  ................
(lldb) 
0x7fffffffd8e8: 40 dd ff ff ff 7f 00 00 7d dd ff ff ff 7f 00 00  @.......}.......
0x7fffffffd8f8: 81 e5 ff ff ff 7f 00 00 9d e5 ff ff ff 7f 00 00  ................
(lldb) memory read argv
0x7fffffffd8a8: 49 dc ff ff ff 7f 00 00 00 00 00 00 00 00 00 00  I...............
0x7fffffffd8b8: 6c dc ff ff ff 7f 00 00 ae dc ff ff ff 7f 00 00  l...............
```

Storing (and repeating) the command string in lldb-dap would break this behavior. What we'd ideally want is to actually take the empty string and pass it to lldb's command interpreter so that the proper repeat logic kicks in.

The thing which makes this tricky (but not too complicated I think) is that lldb-dap multiplexes expression commands and CLI commands into the same string (the `DetectExpressionContext` does the demultiplexing).

I think the proper repeat handling could be two things about each command:
- the type ("expression context") of the command
- the command string itself, if the command was not a CLI command (so we can repeat the expression)

Then, when we get an empty string, we check the type of the previous command:
- if it was a CLI command (`ExpressionContext::Command`), we change send the empty string to lldb command interpreter, so that it does the right thing
- otherwise, we take the expression string and re-evaluate it (like you do here).

ashgti commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -1364,6 +1364,13 @@ void request_evaluate(const llvm::json::Object &request) {
+  // the current expression is empty (i.e. the user hit plain 'return').
+  if (!expression.empty())
+    g_dap.last_nonempty_expression = expression;
+  else
```
Should this be done using https://github.com/llvm/llvm-project/blob/ea2da571c761066542f8d2273933d2523279e631/lldb/include/lldb/API/SBCommandInterpreterRunOptions.h#L77-L81 on the https://github.com/llvm/llvm-project/blob/ea2da571c761066542f8d2273933d2523279e631/lldb/tools/lldb-dap/LLDBUtils.cpp#L48C14-L48C27 call?

cmtice replied:
The SBCommandIntepreterRunOptions don't seem to be used on this execution path, so I couldn't make use of this.

walter-erquinigo commented on lldb/test/API/tools/lldb-dap/evaluate/TestDAP_evaluate.py:
```diff
@@ -60,7 +60,10 @@ def run_test_evaluate_expressions(
 
         # Expressions at breakpoint 1, which is in main
         self.assertEvaluate("var1", "20")
+        # Empty expression should equate to the previous expression.
```
Please add a test that uses `memory read` to make sure that two different memory read invocations give different valid results

cmtice replied:
Done

walter-erquinigo reviewed (changes requested):
Thanks for working on this!
 I second all that Pavel said.

cmtice committed:
[lldb-dap] Add feature to remember last non-empty expression

Update to handle commands & variables separately: empty command
expressions are passed to the CommandIntepreter to handle as it
normally does; empty variable expressions are updated to use the last
non-empty variable expression, if the last expression was a variable (not
a command).  Also updated the test case to test these cases properly, and
added a 'memory read' followed by an empty expression, to make sure it
handles that sequence correctly.

cmtice commented:
> I wrote a fairly long comment on Friday, but I don't see it anymore so, it looks like github has swallowed it. Here's my reconstruction of it:
> 
> LLDB commands have the notion of a "repeat command", which can sometimes be more complicated than just running the same string over and over again. This can be e.g. seen with the `memory read` command, which doesn't just print the same memory again -- it actually prints the memory that comes _after_ it (a pretty nifty feature actually):
> 
> ```
> (lldb) memory read argv
> 0x7fffffffd8a8: 49 dc ff ff ff 7f 00 00 00 00 00 00 00 00 00 00  I...............
> 0x7fffffffd8b8: 6c dc ff ff ff 7f 00 00 ae dc ff ff ff 7f 00 00  l...............
> (lldb) 
> 0x7fffffffd8c8: c4 dc ff ff ff 7f 00 00 cf dc ff ff ff 7f 00 00  ................
> 0x7fffffffd8d8: 09 dd ff ff ff 7f 00 00 18 dd ff ff ff 7f 00 00  ................
> (lldb) 
> 0x7fffffffd8e8: 40 dd ff ff ff 7f 00 00 7d dd ff ff ff 7f 00 00  @.......}.......
> 0x7fffffffd8f8: 81 e5 ff ff ff 7f 00 00 9d e5 ff ff ff 7f 00 00  ................
> (lldb) memory read argv
> 0x7fffffffd8a8: 49 dc ff ff ff 7f 00 00 00 00 00 00 00 00 00 00  I...............
> 0x7fffffffd8b8: 6c dc ff ff ff 7f 00 00 ae dc ff ff ff 7f 00 00  l...............
> ```
> 
> Storing (and repeating) the command string in lldb-dap would break this behavior. What we'd ideally want is to actually take the empty string and pass it to lldb's command interpreter so that the proper repeat logic kicks in.
> 
> The thing which makes this tricky (but not too complicated I think) is that lldb-dap multiplexes expression commands and CLI commands into the same string (the `DetectExpressionContext` does the demultiplexing).
> 
> I think the proper repeat handling could be two things about each command:
> 
> * the type ("expression context") of the command
> * the command string itself, if the command was not a CLI command (so we can repeat the expression)
> 
> Then, when we get an empty string, we check the type of the previous command:
> 
> * if it was a CLI command (`ExpressionContext::Command`), we change send the empty string to lldb command interpreter, so that it does the right thing
> * otherwise, we take the expression string and re-evaluate it (like you do here).

I think I have done what you requested now.

cmtice commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -1363,6 +1363,14 @@ void request_evaluate(const llvm::json::Object &request) {
+  else
+    expression = last_nonempty_expression;
 
   if (context == "repl" && g_dap.DetectExpressionContext(frame, expression) ==
```
Done.

cmtice commented:
I think I have addressed all the review comments; please take another look.

walter-erquinigo commented on lldb/test/API/tools/lldb-dap/evaluate/TestDAP_evaluate.py:
```diff
@@ -191,6 +198,14 @@ def run_test_evaluate_expressions(
         self.assertEvaluate("my_bool_vec", "size=2")
 
+        # Test memory read, especially with 'empty' repeat commands.
+        if context == "repl":
+          self.continue_to_next_stop()
+          self.assertEvaluate("memory read &my_ints",
+                              ".*00 00 00 00 02 00 00 00 04 00 00 00 06 00 00 00.*\n.*08 00 00 00 0a 00 00 00 0c 00 00 00 0e 00 00 00.*\n")
+          self.assertEvaluate("",
+                              ".*10 00 00 00 12 00 00 00 14 00 00 00 16 00 00 00.*\n.*18 00 00 00 1a 00 00 00 1c 00 00 00.*\n")
```
I don't think that this will reliably be the same across systems. Something that would be better would be to read from an `uint8_t` array and just read one byte at a time. That should work anywhere.

cmtice replied:
Done.

walter-erquinigo commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -1364,8 +1364,20 @@ void request_evaluate(const llvm::json::Object &request) {
+       (expression.empty() &&
+        g_dap.last_expression_context == ExpressionContext::Command))) {
+    // If the current expression is empty, and the last expression context was
+    // for a  command, pass the empty expression along to the
```
```suggestion
    // for a command, pass the empty expression along to the
```

cmtice replied:
Done.

walter-erquinigo commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -1364,8 +1364,20 @@ void request_evaluate(const llvm::json::Object &request) {
+    // for a  command, pass the empty expression along to the
+    // CommandInterpreter, to repeat the previous command. Also set the
+    // expression context properly for the next (possibly empty) expression.
+    g_dap.last_expression_context = ExpressionContext::Command;
+    // Since the current expression context is not for a variable, clear the
+    // last_nonempty_var_expression field.
+    g_dap.last_nonempty_var_expression.clear();
```
It would be cleaner if this gets stored in a sort of discriminated union:

LastExpressionInfo:
  - Case Command (no metadata)  
  - Case Var (expression)

this way it's easier to maintain these two variables correctly in sync

cmtice replied:
I don't think a union will work because last_expression_context is needed and used in all empty expression cases, while for the variable case we ALSO need the last_nonempty_var_expression.

cmtice replied:
Actually I have an idea for a way to make this work with only one variable; let me try that.

cmtice replied:
The idea worked; I've updated the code now.

cmtice committed:
[lldb-dap] Add feature to remember last non-empty expression.

Update test to use single-byte ints for memory reads, and to read
one byte at a time. Also fix space in comment.

cmtice committed:
[lldb-dap] Add feature to remember last non-empty expression.

Update to need only one new additional field in DAP structure,
instead of two.

labath commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -1364,8 +1364,14 @@ void request_evaluate(const llvm::json::Object &request) {
-  if (context == "repl" && g_dap.DetectExpressionContext(frame, expression) ==
-                               ExpressionContext::Command) {
+  if (context == "repl" &&
+      ((!expression.empty() &&
+       g_dap.DetectExpressionContext(frame, expression) ==
+       ExpressionContext::Command) ||
+       (expression.empty() && g_dap.last_nonempty_var_expression.empty()))) {
```
```suggestion
      (expression.empty() ? g_dap.last_nonempty_var_expression.empty() :        g_dap.DetectExpressionContext(frame, expression) ==
       ExpressionContext::Command)) {
```

cmtice replied:
Thank you for the suggestion! Done.

walter-erquinigo commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -1376,6 +1382,16 @@ void request_evaluate(const llvm::json::Object &request) {
     EmplaceSafeString(body, "result", result);
     body.try_emplace("variablesReference", (int64_t)0);
   } else {
+    if (context != "hover") {
```
yeah, this feature should be limited to the repl Debug Console

cmtice replied:
Done.

cmtice committed:
[lldb-dap] Add feature to remember last non-empty expression

Clean up formmatting for parameter comment; simplify conditional as
suggested by reviewer; only handle empty variable expressions in 'repl'
context.

cmtice commented on lldb/tools/lldb-dap/LLDBUtils.cpp:
```diff
@@ -45,7 +45,8 @@ bool RunLLDBCommands(llvm::StringRef prefix,
       std::lock_guard<std::mutex> locker(handle_command_mutex);
-      interp.HandleCommand(command.str().c_str(), result);
+      interp.HandleCommand(command.str().c_str(), result,
+                           /* add_to_history */ true);
```
Done.

labath replied:
Okay, I see now. The comment threw me off, as I though this code handles commands which run during termination.

cmtice commented:
All reviewer comments have been addressed. Please take another look. Thank you!

walter-erquinigo commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -1364,8 +1364,14 @@ void request_evaluate(const llvm::json::Object &request) {
-  if (context == "repl" && g_dap.DetectExpressionContext(frame, expression) ==
-                               ExpressionContext::Command) {
+  if (context == "repl" &&
+      (expression.empty() ?
+       g_dap.last_nonempty_var_expression.empty() :
+       g_dap.DetectExpressionContext(frame, expression) ==
```
it's hard to read this. Please move this ternary check into another variable with a good name.

labath replied:
.. but note that the work that DetectExpressionContext does (looking up local variables) is relatively expensive, so we probably don't want to do it when it's not needed (when `context!="repl"`).

cmtice replied:
I've done the best compromise I can -- I've created a new variable, 'repeat_last_command' to use in the if-condition, and tried to set up the if-condition to avoid unnecessary calls to DetectExpressionContext.

walter-erquinigo commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -1376,6 +1382,16 @@ void request_evaluate(const llvm::json::Object &request) {
     body.try_emplace("variablesReference", (int64_t)0);
   } else {
+    if (context == "repl") {
+      // If the expression is empty and the last expression was for a
+      // variable, set the expression to the previous expression (repeat the
+      // evaluation); otherwise save the current non-empty expression for the
+      // next (possibly empty) variable expression.
+      if (expression.empty())
+        expression = g_dap.last_nonempty_var_expression;
```
IIUC, this has the following flow in the debug console:

- You send a command (non-var expression) FOO
- You send a var expression VAR
- You send empty text -> then FOO repeats

If that understanding is correct, that would lead to a confusing behavior. I think it's better to repeat only if the previous input was exactly a command.

cmtice replied:
Your description of the behavior is not correct.   last_nonempty_var_expression is now being used for two purposes: It both holds the last non-empty var expression IF-AND-ONLY-IF the last expression was a var expression; or it indicates that the last expression was a command, by being empty itself. Note the very first thing we do, unconditionally, when processing a command is to clear this variable.

When we receive an empty expression, we check to see if the most recent non-empty expression was a command (test to see if last_nonempty_var_expression is empty). If it was a command, then we pass the empty expression to the command interpreter. If the most recent non-empty expression was a variable expression (i.e. last_nonempty_var_expression is NOT empty), then we treat the empty expression as a variable expression, re-evalulating the contents of last_nonempty_var_expression.

cmtice committed:
[lldb-dap] Add feature to remember last non-empty expression

Move ternary expression out of conditional, and use new boolean
variable instead.

cmtice committed:
[lldb-dap] Add feature to remember last non-empty expression.

Fix formatting issues.

labath commented on lldb/test/API/tools/lldb-dap/evaluate/TestDAP_evaluate.py:
```diff
@@ -54,13 +54,20 @@ def run_test_evaluate_expressions(
         self.assertEvaluate("var1", "20")
+        # Empty expression should equate to the previous expression.
+        if context == "repl":
+            self.assertEvaluate("", "20")
```
```suggestion
            self.assertEvaluate("", "20")
        else:
            self.assertEvaluateFailure("")
```
.. or something along those lines (I haven't checked whether this works) -- basically to check that the empty string does not repeat in non-repl contexts.

cmtice replied:
Done

labath reviewed (commented):
This works for me. Walter?

cmtice committed:
[lldb-dap] Add feature to remember last non-empty expression

Update test slightly (verfiy empty variable expression fails in
non-repl mode).

walter-erquinigo reviewed (approved):
lgtm!
